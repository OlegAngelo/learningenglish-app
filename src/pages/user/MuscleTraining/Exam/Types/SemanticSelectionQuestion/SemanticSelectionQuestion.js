import React, { Fragment, useEffect, useContext } from 'react';

import styles from './SemanticSelectionQuestion.module.css';

import QuestionnaireItem from '../../components/QuestionnaireItem';
import SpeakerIcon from '../../../../../../shared/icons/SpeakerIcon';

import questionHelper from '../../../../../../utils/questionHelper';
import QuestionSpeakerPlayer from '../../components/QuestionSpeakerPlayer/QuestionSpeakerPlayer';
import { QuestionWrapperContext } from '../../components/QuestionWrapper/QuestionWrapper.js';

const SemanticSelectionQuestion = () => {
  const {
    timerProps,
    footerProps,
    setFooterProps,
    isShowCommentary,
    response,
    setResponse,
    selected,
    setSelected,
    questionItem,
    setAnswers,
    checkAnswer,
    resultMessage,
    categories,
    AudioErrorModal,
    play,
    isAudioPlaying,
  } = useContext(QuestionWrapperContext);
  const {
    title,
    description,
    choices,
    id,
    audio_file: audioFile,
    pronunciation_point: pronuncationTips,
  } = questionItem;
  const correctChoice = choices.filter(item => item.is_correct)[0].jp_item;
  const learningType = choices[0];
  const parsedTitle = title.replace(/#★|☆#/g, '');

  const handleAnswerOptionClick = (choice) => {
    if (selected) return;
    setSelected(choice.jp_item);
    setResponse(checkAnswer(choice.is_correct, timerProps.seconds) ? resultMessage : 'Failed...');
    setAnswers((answers) => [
      ...answers,
      questionHelper.generateQuestionResult(categories, id, choice.id),
    ]);
  };

  const sessionQuestion = () => {
    if (learningType.training_word_id) {
      return "単語の意味を選択しましょう";
    } else {
      return "英語の意味を選択しましょう";
    }
  };

  useEffect(() => {
    setFooterProps({
      ...footerProps,
      hasSpeaker: false,
      hasNext: response && true,
      hasLightBulb: false,
      hasSkip: !response,
    });
  }, [response]);

  return (
    <div className="w-full">
      <AudioErrorModal />
      <p className="mx-2 font-semibold text-16 leading-px-23 text-center text-primary-500">{!response && sessionQuestion()}</p>
      <p className={`mx-6 text-basic-100 ${parsedTitle.length > 50 && !isShowCommentary ? `leading-px-29 mt-px-15 text-20` : `${styles.englishQuestion} mt-px-34 text-24`}`}>{parsedTitle}</p>
      <p className={`${styles.correctChoice} mx-6 text-basic-100 ${parsedTitle.length > 50 && response ? `text-14` : `text-18`}`}>{response && correctChoice}</p>
      {isShowCommentary && audioFile ? (
        <div className="pt-px-21 flex justify-center">
          <button
            onClick={() => play(categories)}
            disabled={isAudioPlaying}
            className="disabled:opacity-50"
          >
            <SpeakerIcon width="25" height="24" className={`text-center`}/>
            <div className={`text-8 text-center text-primary-400 font-bold mt-px-4`}>
              問題を再生
            </div>
          </button>
        </div>
        ) : (
          <div>
            {response && audioFile && (
              <div>
                <QuestionSpeakerPlayer
                  className="my-px-16 w-full"
                  category={categories}
                  isAudioPlaying={isAudioPlaying}
                  play={play}
                  key={questionItem.id}
                  response={response}
                />
              </div>
            )}
          </div>
        )
      }

      <div className={`mx-2 mb-2 ${styles.contentContainer}`}>
        {isShowCommentary
          ? (
            description &&
            <Fragment>
              <div className="mt-px-38 mb-px-8 p-4 rounded bg-basic-400">
                <p className={`mb-2 font-bold text-16 ${styles.commentaryTitle} text-primary-500`}>解説</p>
                <p className={`mb-px-18 text-14 ${styles.commentaryDescription} whitespace-pre-wrap ${styles.textFormat} text-basic-100`}>{description}</p>
              </div>
            </Fragment>
          ) : (
            <Fragment>
              <div
                className={`h-px-38 text-24 leading-px-38 text-center font-black ${styles.response}`}
                style={{ color: questionHelper.getResponseColor(response) }}
              >
                &nbsp;{response}
              </div>
              {choices.map((choice, key) => (
                <QuestionnaireItem
                  key={key}
                  text={choice.jp_item}
                  onClick={() => handleAnswerOptionClick(choice)}
                  color={
                    selected && choice.is_correct
                      ? 'bg-secondary-500'
                      : 'bg-basic-400'
                  }
                  textColor={questionHelper.isIncorrectChoice(choice.jp_item, selected, choice.is_correct) ? '#E34E42' : '#141414'}
                />
              ))}
            </Fragment>
          )
        }
      </div>
    </div>
  );
};

export default SemanticSelectionQuestion;
