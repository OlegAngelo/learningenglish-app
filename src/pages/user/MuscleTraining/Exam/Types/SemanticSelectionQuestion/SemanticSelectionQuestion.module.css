.response {
  margin-bottom: -8px;
  -webkit-text-stroke: .55px #FFFFFF;
}

.commentaryTitle {
  line-height: 23.2px;
}

.commentaryDescription {
  line-height: 23.8px;
}

.englishQuestion {
  line-height: 35.52px;
}

.correctChoice {
  line-height: 26.64px;
  margin-top: 8px;
}

.contentContainer {
  padding-top: 65px;
}

.textFormat {
  line-break: strict;
  word-wrap: break-word;
  overflow-wrap: break-word;
}

/* ----------- iPhone 6+, 7+ and 8+ ----------- */
@media only screen
  and (max-device-width: 414px)
  and (min-device-height: 736px)
  and (-webkit-min-device-pixel-ratio: 3)
  and (orientation: portrait) {

  .contentContainer {
    padding-top: 40px;
  }

}

/* ----------- iPhone 6, 6S, 7 and 8 ----------- */
@media only screen
  and (max-device-width: 375px)
  and (min-device-height: 667px)
  and (-webkit-min-device-pixel-ratio: 2)
  and (orientation: portrait) {

  .contentContainer {
    padding-top: 10px;
  }

}

/* ----------- iPhone X ----------- */
@media only screen
  and (max-device-width: 375px)
  and (min-device-height: 812px)
  and (-webkit-min-device-pixel-ratio: 3)
  and (orientation: portrait) {

  .contentContainer {
    padding-top: 65px;
  }

}