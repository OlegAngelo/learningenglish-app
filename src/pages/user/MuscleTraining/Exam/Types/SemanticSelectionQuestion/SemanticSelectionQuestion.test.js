import 'jsdom-global/register';
import React, { useState as useStateMock } from 'react';
import { configure, shallow, mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

import SemanticSelectionQuestion from './SemanticSelectionQuestion';

configure({ adapter: new Adapter() });

describe('SemanticSelectionQuestion Component', () => {
  let props = {
    footerProps: {
      hasLightBulb: false,
      hasBack: false,
      hasSkip: false,
      hasSpeaker: true,
      hasNext: true,
    },
    questionItem: {
      id: 'ED1_P0001',
      title: 'Let’s start the #★meeting☆#.',
      jp_title: '会議を始めましょう。',
      audio_file: 'ED1_P0001.mp3',
      choices: [
        {
          id: 1,
          phrase_item: 'Let’s start the meeting.',
          word_item: 'meeting',
          jp_phrase_item: '会議を始めましょう。',
          is_correct: true,
        },
        {
          id: 2,
          phrase_item: 'Let’s start the chat.',
          word_item: 'mapping',
          jp_phrase_item: '文通を始めましょう。',
          is_correct: false,
        },
        {
          id: 3,
          phrase_item: 'Let’s start the project.',
          word_item: 'shooting',
          jp_phrase_item: '面談日を決めましょう。',
          is_correct: false,
        },
        {
          id: 4,
          phrase_item: 'Let’s start the exhibition.',
          word_item: 'viewing',
          jp_phrase_item: '録音を開始しましょう。',
          is_correct: false,
        },
      ],
      typing: 'meeting',
      description: '会議の開始のときによく使うセリフです。Let\'s start off the meeting now.、Let\'s get started.などの言い方もします。',
      pronunciation_point: 'Let\'s startを「レッスタート」のように一気に言います。',
    },
    selected: "会議を始めましょう。",
    setFooterProps: jest.fn(),
    response: '',
    setResponse: jest.fn(),
    selected: '',
    setSelected: jest.fn(),
    isShowCommentary: false
  };

  describe('render', () => {
    it('should render exaclty from snapshot', () => {
      const wrapper = mount(<SemanticSelectionQuestion {...props} />)
      expect(wrapper).toMatchSnapshot()
    });
  });

  describe('props', () => {
    it('should pass down to component', () => {
      const wrapper = mount(<SemanticSelectionQuestion {...props} />);
      const wrapperProps = wrapper.props();

      expect(wrapperProps.footerProps).toEqual(props.footerProps);
      expect(wrapperProps.setFooterProps).toEqual(props.setFooterProps);
      expect(wrapperProps.response).toEqual(props.response);
      expect(wrapperProps.setResponse).toEqual(props.setResponse);
      expect(wrapperProps.selected).toEqual(props.selected);
      expect(wrapperProps.setSelected).toEqual(props.setSelected);
      expect(wrapperProps.isShowCommentary).toEqual(props.isShowCommentary);
    });
  });

  describe("Questionnaire Title Text", () => {
    it('should check if the questionnaire item title is similar to the dynamic data', () => {
      const wrapper = mount(<SemanticSelectionQuestion {...props} />);
      const QuestionTitle = props.questionItem.title.replace(/#★|☆#/g, "");
      expect(wrapper.find('.englishQuestion').text()).toContain(QuestionTitle);
    });
  });

  describe("Questionnaire Items text", () => {
    it('should check if the question item choices are same from dynamic data', () => {
      const wrapper = mount(<SemanticSelectionQuestion {...props} />);

      expect(wrapper.find('button').contains(props.questionItem.choices[0].jp_phrase_item));
      expect(wrapper.find('button').contains(props.questionItem.choices[1].jp_phrase_item));
      expect(wrapper.find('button').contains(props.questionItem.choices[2].jp_phrase_item));
      expect(wrapper.find('button').contains(props.questionItem.choices[3].jp_phrase_item));
    });
  });

});
