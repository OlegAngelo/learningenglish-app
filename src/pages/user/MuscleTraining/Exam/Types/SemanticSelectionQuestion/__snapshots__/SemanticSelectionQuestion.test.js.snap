// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SemanticSelectionQuestion Component render should render exaclty from snapshot 1`] = `ReactWrapper {}`;
